### PR TITLE
subdivided vision cones

### PR DESCRIPTION
**Pull request by Koppeprojects**
Fix #

#### Summary of additions and changes

* 

#### How to test this pull request

```

```

### CODE REVIEW

File: swarmrl/observables/subdivided_vision_cones.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                Normalised director of the colloid.
+        colloids : List[Colloid]
+                all colloids
+        Returns
+        np.array of shape (n_cones, num_of_types) containing the vision values
+        for each cone and for each particle type that can be visible.
+        """
+        vision_val_out = np.zeros((self.n_cones, len(self.types)))
+
+        for c in colloids:
```
Comment (SamTov) [inline swarmrl/observables/subdivided_vision_cones.py]: We could probably parallelised this using `jax.vmap` but perhaps this can be looked at later as it is not critical at the moment.

File: swarmrl/observables/subdivided_vision_cones.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        ----------
+        colloids : object
+                Colloids with all possible different types.
+
+        Returns
+        all possible types in the corresponding index they will also later be found in,
+        when the observable is calculated
+        """
+        types = []
+        for c in colloids:
```
Comment (SamTov) [inline swarmrl/observables/subdivided_vision_cones.py:72]: How do we handle types in the vision cone? Should we return a vector or a weight or something?

File: swarmrl/observables/subdivided_vision_cones.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+from typing import List
+
+import numpy as np
+
+from swarmrl.models.interaction_model import Colloid
+from swarmrl.observables.observable import Observable
+from swarmrl.tasks.task import Task
+
+
+class SubdividedVisionCones(Observable, ABC, Task):
```
Comment (SamTov) [inline swarmrl/observables/subdivided_vision_cones.py]: Why is this a child class of Task? I think it should only be observable.

**Review by SamTov — CHANGES_REQUESTED** [review]
Looks really good! All the doc strings are perfect and the test looks good. Please address the comments in the review and adapt the parent classes.

File: CI/unit_tests/observables/test_subdivided_vision_cone.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            n_cones=3,
+            radii=[1, 2, 3, 4, 1],
+        )
+
+    def test_call(self):
+        """
+        Tests if the computation of the observable is still correct.
+        """
+
+        class Colloid:
```
Comment (christophlohrmann) [inline CI/unit_tests/observables/test_subdivided_vision_cone.py]: do we need this mock colloid or can we use the actual ``Colloid`` class that is used in the package.
If we do need a mock colloid, move it to a separate file so all tests can use it

File: swarmrl/observables/subdivided_vision_cones.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    The vision cone acts like a camera for the particles. It can either output all
+    colloids within its view, a function of the distances between one colloid and all
+    other colloids, or the normalised distance to the source.
+    """
+
+    def __init__(
+        self, vision_range: float, vision_half_angle: float, n_cones: int, radii=[]
```
Comment (christophlohrmann) [inline swarmrl/observables/subdivided_vision_cones.py]: is radii=[] a case of mutable default argument? If yes replace by radii:typing.List[float] = None and then do the None check below

File: swarmrl/observables/subdivided_vision_cones.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                List of the radii of the colloids in the experiment,
+                ordered by the ids of the colloids
+        """
+        self.vision_range = vision_range
+        self.vision_half_angle = vision_half_angle
+        self.n_cones = n_cones
+        self.radii = radii
+        self.types = []
+
+    def initialize(self, colloids: List[Colloid]):
```
Comment (christophlohrmann) [inline swarmrl/observables/subdivided_vision_cones.py]: if it does nothing (and the parent class fcn does also just pass) omit this function

File: swarmrl/observables/subdivided_vision_cones.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    def detect_things_to_see(self, colloids: List[Colloid]):
+        """
+        Calculates the normalised director of the colloids.
+
+        Parameters
+        ----------
+        colloids : object
+                Colloids with all possible different types.
+
+        Returns
```
Comment (christophlohrmann) [inline swarmrl/observables/subdivided_vision_cones.py:66]: actually resturns nothing, but updates the class member

Comment (christophlohrmann) [reply swarmrl/observables/subdivided_vision_cones.py:66]: also consider prepending an underscore for this private method

File: swarmrl/observables/subdivided_vision_cones.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        ----------
+        colloid : object
+                Colloid for which the observable should be computed.
+
+        Returns
+        Colloid position and vision cone director.
+        """
+        my_pos = np.copy(colloid.pos[:2])
+        my_director = np.copy(colloid.director[:2])
+        # TODO: Add 3D support
```
Comment (christophlohrmann) [inline swarmrl/observables/subdivided_vision_cones.py:91]: don't do this 3d-2d conversion. Just go over the directors once and raise an error if pone has a z-component != 0

File: swarmrl/observables/subdivided_vision_cones.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            # make sure not to see yourself ;
+            # put vision_range sufficiently high if no upper limit is wished
+            if dist_norm == 0 or not in_range:
+                continue
+            # calc perceived angle deviation ( sign of angle is missing )
+
+            angle = np.arccos(np.dot(dist / dist_norm, my_director))
+            # use the director in orthogonal direction to determine sign
+            orthogonal_dot = np.dot(dist / dist_norm, [-my_director[1], my_director[0]])
+            angle *= np.sign(orthogonal_dot)
```
Comment (christophlohrmann) [inline swarmrl/observables/subdivided_vision_cones.py]: consider moving this signed-angle computation to its own file (in utils somewhere) and testing it separately

File: swarmrl/observables/subdivided_vision_cones.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+            for cone in range(self.n_cones):
```
Comment (christophlohrmann) [inline swarmrl/observables/subdivided_vision_cones.py]: rename cone->cone_idx to help understand that it is an int

**Review by christophlohrmann — CHANGES_REQUESTED** [review]
overall pretty good, thanks for putting in the work. @SamTov will the number-of-types-dependent output shape be a problem for the ML models?

**Comment by SamTov** [comment]
> overall pretty good, thanks for putting in the work. @SamTov will the number-of-types-dependent output shape be a problem for the ML models?

Maybe we can have the option to combine it into a single cone or to pass back seperate cones because it will change our model definitions.

File: swarmrl/utils/utils.py [hunk 11]
```diff
@@ -271,3 +272,36 @@ def record_training(training_dict: dict):
...
+            Normalised director in 3D.
+    other_director : np.ndarray
+            Normalised director in 3D.
+    Returns
+    singed float which represents the signed angle of my_director to otherdirector
+    with the mathematical sign convention.
+    """
+
+    # Assert if the directors were really normalized
+    # np.testing.assert_almost_equal(jnp.linalg.norm(my_director), 1, decimal=6)
```
Comment (SamTov) [inline swarmrl/utils/utils.py]: The assertions are a nice idea, but instead of raising an error, maybe try the following:

```python
if np.linalg.norm(my_director) != 1:
    my_director /= np.linalg.norm(my_director)
```
In the case where you cannot have an if statement here, e.g. traced jax code, just perform the normalization as the overhead such as:

```python
my_director /= np.linalg.norm(my_director)
```

File: swarmrl/utils/utils.py [hunk 12]
```diff
@@ -271,3 +272,36 @@ def record_training(training_dict: dict):
...
+
+    Parameters
+    ----------
+
+    my_director : np.ndarray
+            Normalised director in 3D.
+    other_director : np.ndarray
+            Normalised director in 3D.
+    Returns
+    singed float which represents the signed angle of my_director to otherdirector
```
Comment (SamTov) [inline swarmrl/utils/utils.py]: Update the doc string to:

```python
def calc_signed_angle_between_directors(
    my_director: np.ndarray, other_director: np.ndarray
) -> float:
    """
    In 2D compare two different normalized
    directors to determine the angle between them

    Parameters
    ----------
    my_director : np.ndarray
            Normalised director in 3D.
    other_director : np.ndarray
            Normalised director in 3D.
    Returns
    -------
    signed_angle : float
            Signed float which represents the signed angle of my_director to other director
    """
```
There is also a spelling mistake for singed - signed

File: CI/unit_tests/utils/test_utils.py [hunk 13]
```diff
@@ -68,3 +71,30 @@ def test_gather_n_dim_indices(self):
         )
         test_selection = gather_n_dim_indices(data, indices)
         assert_array_equal(test_selection, true_selection)
+
+    def test_calc_signed_angle_between_directors(self):
+        """
+        Test the calc_signed_angle_between_directors
+        for positiv and negativ angles in 2D
```
Comment (SamTov) [inline CI/unit_tests/utils/test_utils.py:78]: Add a test checking for the automatic normalization, e.g:

```python
my_vector = np.array([3, 2, 0])
my_normed_vector = my_vector / np.linalg.norm(my_vector)

assert calc_signed_angle_between_directors(my_vector) == calc_signed_angle_between_directors(my_normed_vector)
```

File: swarmrl/observables/subdivided_vision_cones.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+        my_director = my_director / np.linalg.norm(my_director)
+        assert abs(colloid.pos[2]) < 10e-6
+        assert abs(colloid.director[2]) < 10e-6
+        return my_pos, my_director
+
+    def _calculate_cones_single_object(
+        self, my_pos, my_director, c_type, c_pos, radius
+    ):
+        # generate a blue print of the output values
+        vision_val_out = jnp.ones((self.n_cones, len(self.detected_types)))
```
Comment (SamTov) [inline swarmrl/observables/subdivided_vision_cones.py:126]: If you used `np.zeros` here there would be no memory usage until the array was populated. This is a truly unnecessary thing or us to do I just wanted to point it out.

File: swarmrl/observables/subdivided_vision_cones.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+        # TODO: Add 3D support
+        my_director = my_director / np.linalg.norm(my_director)
+        assert abs(colloid.pos[2]) < 10e-6
+        assert abs(colloid.director[2]) < 10e-6
+        return my_pos, my_director
+
+    def _calculate_cones_single_object(
+        self, my_pos, my_director, c_type, c_pos, radius
+    ):
+        # generate a blue print of the output values
```
Comment (SamTov) [inline swarmrl/observables/subdivided_vision_cones.py:125]: Add a doc string and type hints please.

File: swarmrl/observables/subdivided_vision_cones.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+                For example [0,2] here colloids of type 1 won't be detected.
+                if None then al will be detected.
+
+        """
+        self.vision_range = vision_range
+        self.vision_half_angle = vision_half_angle
+        self.n_cones = n_cones
+        self.radii = radii
+        self.detected_types = detected_types
+
```
Comment (SamTov) [inline swarmrl/observables/subdivided_vision_cones.py:56]: you can add a jit compiled version of the angle function in the init:

```python
self.angle_fn = jax.jit(calc_signed_angle_between_directors)
```

later in the code you can then call:

```python
self.angle_fn(my_director, other_director)
```

**Review by SamTov — CHANGES_REQUESTED** [review]
I think it looks nice, I have a few more points that would be nice to see included. Also, Christoph's points still need addressing.

File: CI/unit_tests/utils/test_utils.py [hunk 17]
```diff
@@ -68,3 +71,45 @@ def test_gather_n_dim_indices(self):
...
+            my_director2 * 1.5, other_director3
+        )
+
+        assert angle1 == 0
+        assert angle2 == np.pi / 2
+        assert angle3 == np.pi / 3
+        assert angle4 == np.pi * 3 / 4
+        assert angle5 == -np.pi * 3 / 4
+        assert abs(angle6 + np.pi * 11 / 12) < 10e-6
+        assert abs(angle7 - angle3) < 10e-6
```
Comment (SamTov) [inline CI/unit_tests/utils/test_utils.py:113]: Just for reference, there are things like `pytest.approx` for this situation just for the future.

**Review by SamTov — DISMISSED** [review]
It is good on my end. Christoph still needs to sign off but I think you need approval from us both so I will add my approval but this shouldn't be merged until @christophlohrmann is also happy with the changes.

File: CI/unit_tests/utils/test_utils.py (hunk 17 above)
Comment (christophlohrmann) [reply CI/unit_tests/utils/test_utils.py:113]: np.testing.assert_allclose() is also fantastic because you can set absolute and relative tolerance independently and it works with scalars and arrays

File: swarmrl/utils/utils.py (hunk 11 above)
Comment (christophlohrmann) [reply swarmrl/utils/utils.py]: I disagree. we should just check that the system is 2d because the observable only works in 2d. directors never need to be normalized because they are by default

**Review by christophlohrmann — DISMISSED** [review]

**Comment by christophlohrmann** [comment]
good to go from my side

**Review by christophlohrmann — APPROVED** [review]

File: swarmrl/utils/utils.py (hunk 11 above)
Comment (SamTov) [reply swarmrl/utils/utils.py]: If the director is normalised by default then sure, there is no need for the assertion or a default normalisation procedure and we can avoid the problem. Happy to remove that.

File: CI/unit_tests/utils/test_utils.py (hunk 17 above)
Comment (SamTov) [reply CI/unit_tests/utils/test_utils.py:113]: Fully agree!
